services/kvstore/src/main/java/tech/pegasys/pantheon/services/kvstore/InMemoryKeyValueStorage.java

```diff
@@ -49,71 +52,65 @@ public void clear() {
   }
 
   @Override
-  public void close() {}
-
-  @Override
-  public boolean containsKey(final BytesValue key) throws StorageException {
+  public boolean containsKey(final byte[] key) throws StorageException {
     final Lock lock = rwLock.readLock();
     lock.lock();
     try {
-      return hashValueStore.containsKey(key);
+      return hashValueStore.containsKey(BytesValue.wrap(key));
     } finally {
       lock.unlock();
     }
   }
 
   @Override
-  public Optional<BytesValue> get(final BytesValue key) {
+  public Optional<byte[]> get(final byte[] key) throws StorageException {
     final Lock lock = rwLock.readLock();
     lock.lock();
     try {
-      return Optional.ofNullable(hashValueStore.get(key));
+      return Optional.ofNullable(hashValueStore.get(BytesValue.wrap(key)));
     } finally {
       lock.unlock();
     }
   }
 
   @Override
-  public long removeUnless(final Predicate<BytesValue> inUseCheck) {
-    final Lock lock = rwLock.writeLock();
-    lock.lock();
-    try {
-      long initialSize = hashValueStore.keySet().size();
-      hashValueStore.keySet().removeIf(key -> !inUseCheck.test(key));
-      return initialSize - hashValueStore.keySet().size();
-    } finally {
-      lock.unlock();
-    }
+  public long removeAllKeysUnless(final Predicate<byte[]> retainCondition) throws StorageException {
+    long initialSize = hashValueStore.keySet().size();
+    hashValueStore.keySet().removeIf(key -> !retainCondition.test(key.getArrayUnsafe()));
+    return initialSize - hashValueStore.keySet().size();
   }
 
   @Override
-  public Transaction startTransaction() {
-    return new InMemoryTransaction();
+  public void close() {}
+
+  @Override
+  public KeyValueStorageTransaction startTransaction() {
+    return new KeyValueStorageTransactionTransitionValidatorDecorator(new InMemoryTransaction());
   }
 
   public Set<BytesValue> keySet() {
     return Set.copyOf(hashValueStore.keySet());
   }
 
-  private class InMemoryTransaction extends AbstractTransaction {
+  private class InMemoryTransaction implements KeyValueStorageTransaction {
 
-    private Map<BytesValue, BytesValue> updatedValues = new HashMap<>();
+    private Map<BytesValue, byte[]> updatedValues = new HashMap<>();
     private Set<BytesValue> removedKeys = new HashSet<>();
 
     @Override
-    protected void doPut(final BytesValue key, final BytesValue value) {
-      updatedValues.put(key, value);
-      removedKeys.remove(key);
+    public void put(final byte[] key, final byte[] value) {
+      updatedValues.put(BytesValue.wrap(key), value);
+      removedKeys.remove(BytesValue.wrap(key));
     }
 
     @Override
-    protected void doRemove(final BytesValue key) {
-      removedKeys.add(key);
-      updatedValues.remove(key);
+    public void remove(final byte[] key) {
+      removedKeys.add(BytesValue.wrap(key));
```

Comment:
do we need to do any null or empty array check?

Reply:
No, this class falls within the boundary of trust, i.e. not expecting a `null` input as Pantheon is assumed to be a responsible actor

Reply:
👍